File: projects/native-apps/ios/Shared/UploadJobStore.swift
import Foundation

struct UploadJobStore {
	private let directory: URL

	init(containerURL: URL) {
		directory = containerURL.appendingPathComponent("Library/Application Support/upload-queue", isDirectory: true)
	}

	static func inSharedContainer(appGroupId: String) -> UploadJobStore? {
		FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId).map(UploadJobStore.init(containerURL:))
	}

	func bytesURL(for job: UploadJob) -> URL {
		directory.appendingPathComponent("\(job.id).multipart")
	}

	func admit(_ job: UploadJob) async throws {
		try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
		for superseded in decodedRecords() where superseded.url == job.url {
			remove(superseded)
		}
		try write(job)
	}

	func stageReady(_ job: UploadJob, form: MultipartForm) async throws -> UploadJob {
		try form.write(to: bytesURL(for: job))
		let ready = job.staged(contentType: form.contentType)
		try write(ready)
		return ready
	}

	func update(_ job: UploadJob) throws {
		try write(job)
	}

	func loadAll(now: Date) -> [UploadJob] {
		decodedRecords()
			.filter { $0.isDue(now: now) }
			.sorted { $0.createdAt < $1.createdAt }
	}

	func remove(_ job: UploadJob) {
		try? FileManager.default.removeItem(at: recordURL(for: job))
		try? FileManager.default.removeItem(at: bytesURL(for: job))
	}

	func removeOrphanedBytes() {
		let recorded = Set(files(withExtension: "json").map { $0.deletingPathExtension().lastPathComponent })
		for body in files(withExtension: "multipart")
		where !recorded.contains(body.deletingPathExtension().lastPathComponent) {
			try? FileManager.default.removeItem(at: body)
		}
	}

	func purgeAll() {
		try? FileManager.default.removeItem(at: directory)
	}

	private func recordURL(for job: UploadJob) -> URL {
		directory.appendingPathComponent("\(job.id).json")
	}

	private func write(_ job: UploadJob) throws {
		try JSONEncoder().encode(job).write(to: recordURL(for: job), options: .atomic)
	}

	private func decodedRecords() -> [UploadJob] {
		let decoder = JSONDecoder()
		return files(withExtension: "json").compactMap { record in
			guard let data = try? Data(contentsOf: record) else { return nil }
			return try? decoder.decode(UploadJob.self, from: data)
		}
	}

	private func files(withExtension pathExtension: String) -> [URL] {
		let entries = (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)) ?? []
		return entries.filter { $0.pathExtension == pathExtension }
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/core/UploadJobStore.kt
package com.readplace.android.core

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerializationException
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.time.Instant

class UploadJobStore(filesRoot: File, private val io: CoroutineDispatcher) : PurgeableUploadQueue {
	private val directory = File(filesRoot, "upload-queue")

	fun bytesFile(job: UploadJob): File = File(directory, "${job.id}.multipart")

	suspend fun admit(job: UploadJob): Unit =
		withContext(io) {
			Files.createDirectories(directory.toPath())
			decodedRecords().filter { it.url == job.url }.forEach(::remove)
			write(job)
		}

	suspend fun stageReady(job: UploadJob, form: MultipartForm): UploadJob =
		withContext(io) {
			form.writeTo(bytesFile(job))
			val ready = job.staged(form.contentType)
			write(ready)
			ready
		}

	fun update(job: UploadJob) {
		write(job)
	}

	fun loadAll(now: Instant): List<UploadJob> =
		decodedRecords()
			.filter { it.isDue(now) }
			.sortedBy { it.createdAt }

	fun remove(job: UploadJob) {
		recordFile(job).delete()
		bytesFile(job).delete()
	}

	fun removeOrphanedBytes() {
		val recorded = files(extension = "json").map { it.nameWithoutExtension }.toSet()
		files(extension = "multipart")
			.filter { it.nameWithoutExtension !in recorded }
			.forEach { it.delete() }
	}

	override fun purgeAll() {
		directory.deleteRecursively()
	}

	private fun recordFile(job: UploadJob): File = File(directory, "${job.id}.json")

	private fun write(job: UploadJob) {
		val staged = File(directory, "${job.id}.json.tmp")
		staged.writeText(job.toRecord())
		Files.move(
			staged.toPath(),
			recordFile(job).toPath(),
			StandardCopyOption.ATOMIC_MOVE,
			StandardCopyOption.REPLACE_EXISTING,
		)
	}

	private fun decodedRecords(): List<UploadJob> = files(extension = "json").mapNotNull(::decodedRecord)

	private fun decodedRecord(record: File): UploadJob? =
		try {
			UploadJob.fromRecord(record.readText())
		} catch (_: IOException) {
			null
		} catch (_: SerializationException) {
			null
		}

	private fun files(extension: String): List<File> =
		directory.listFiles().orEmpty().filter { it.extension == extension }
}
